### PR TITLE
Support Rename refactoring

### PR DESCRIPTION
**Pull request by svor**
Adds a set of commands to support rename refactoring:
- Validate new name;
- Create linked model for rename;
- Recognize type of renamed element to show appropriate wizard;
- Call rename operation.

This PR depends on https://github.com/eclipse/eclipse.jdt.ls/pull/614

### CODE REVIEW

**Comment by fbricon** [comment]
In https://youtu.be/HkCKpGSP05o?t=77 the preview panel as few issues:
- it's missing labels to know which one is before, which one is after changes. Eclipse has Original Source vs Refactored Source labels
- Refactored source panel is on the left, which is counter intuitive, again compare to Eclipse's order
- Refactored source name panel is still named B.java, instead of B1.java (the new name)

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+   * document. Whether a client supports versioned document edits is expressed via
+   * `WorkspaceClientCapabilities.versionedWorkspaceEdit`.
+   */
+  @Pure
+  public List<TextDocumentEdit> getDocumentChanges() {
+    return this.documentChanges;
+  }
+
+  /**
+   * An array of `TextDocumentEdit`s to express changes to specific a specific version of a text
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: Actually I've copied this class from org.eclipse.lsp4j.WorkspaceEdit and made some minor changes which are related to resource changes. I've noticed it in javadoc.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+      return true;
+    }
+    if (obj == null) {
+      return false;
+    }
+    if (getClass() != obj.getClass()) {
+      return false;
+    }
+    CheWorkspaceEdit other = (CheWorkspaceEdit) obj;
+    if (this.changes == null) {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: see the comment above

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedData.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+import java.util.List;
+
+/**
+ * Contains data of linked model.
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedData.java]: I've reworked the model

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModeModel.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * Contributors:
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * The model of the linked mode.
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModeModel.java]: I've reworked the model of linked mode

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModeModel.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+  public List<LinkedPositionGroup> getGroups() {
+    return groups;
+  }
+
+  public void setGroups(List<LinkedPositionGroup> groups) {
+    this.groups = groups;
+  }
+
+  public void addGroup(LinkedPositionGroup group) {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModeModel.java]: I've reworked the model of linked mode

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModelParams.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+/**
+ * Describes parameters for activating linked mode of Rename refactoring.
+ *
+ * @author Valeriy Svydenko
+ */
+public class LinkedModelParams {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModelParams.java]: I've reworked the model of linked mode

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModelParams.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  public String getUri() {
+    return uri;
+  }
+
+  public void setUri(String uri) {
+    this.uri = uri;
+  }
+
+  /** Returns offset of the cursor position. */
+  public int getOffset() {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/LinkedModelParams.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/Region.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+/**
+ * Represents some region(text range) in editor document.
+ *
+ * @author Valeriy Svydenko
+ */
+public class Region {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/Region.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/RenameSettings.java [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+   */
+  public boolean isUpdateTextualMatches() {
+    return updateTextualMatches;
+  }
+
+  public void setUpdateTextualMatches(boolean updateTextualMatches) {
+    this.updateTextualMatches = updateTextualMatches;
+  }
+
+  enum MachStrategy {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/RenameSettings.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/RenameWizardType.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+package org.eclipse.che.jdt.ls.extension.api.dto;
+
+import org.eclipse.che.jdt.ls.extension.api.RenameType;
+
+/**
+ * The type of Rename refactoring.
+ *
+ * @author Valeriy Svydenko
+ */
+public class RenameWizardType {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/RenameWizardType.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/ChangeUtil.java [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+   * @param manager preference manager
+   * @throws CoreException
+   */
+  public static void convertChanges(Change change, CheWorkspaceEdit edit, PreferenceManager manager)
+      throws CoreException {
+    if (change == null || !(change instanceof CompositeChange)) {
+      return;
+    }
+
+    preferenceManager = manager;
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/ChangeUtil.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetLinkedModelCommand.java [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.corext.dom.LinkedNodeFinder;
+import org.eclipse.jdt.ls.core.internal.corext.refactoring.util.RefactoringASTParser;
+
+/**
+ * Command to to prepare linked model for rename refactoring.
+ *
+ * @author Valeriy Svydenko
+ */
+public class GetLinkedModelCommand {
```
[the remark this answers is not in the document]
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetLinkedModelCommand.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+/**
+ * It is temporary replacement of {@link org.eclipse.lsp4j.WorkspaceEdit} for transporting changes
+ * of the refactoring operation.
+ *
+ * <p>A workspace edit represents changes to many resources managed in the workspace. The edit
+ * should either provide `changes` or `documentChanges`. If documentChanges are present they are
+ * preferred over `changes` if the client can handle versioned document edits.
+ */
+@SuppressWarnings("all")
+public class CheWorkspaceEdit {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: As mentioned in the companion PR, we should add a "kind" when creating resources.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetRenamingElementInfoCommand.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+      return result;
+    } catch (CoreException e) {
+      JavaLanguageServerPlugin.logException(e.getMessage(), e);
+    }
+
+    return result;
+  }
+
+  private static RenameKind getWizardType(IJavaElement element) throws CoreException {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetRenamingElementInfoCommand.java]: "getRenameKind()" instead

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java (hunk 13 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: Created a new class `CheResourceChange` which is extended `ResourceChange` and added a kind of the changed resource. Also reworked `CheWorkspaceEdit`. 
@tsmaeder please take a look.

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetRenamingElementInfoCommand.java (hunk 14 above)
Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.core/src/main/java/org/eclipse/che/jdt/ls/extension/core/internal/refactoring/rename/GetRenamingElementInfoCommand.java]: fixed

File: org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java [hunk 15]
```diff
@@ -10,162 +10,28 @@
...
-  private Map<String, List<TextEdit>> changes;
-
-  /**
-   * An array of `TextDocumentEdit`s to express changes to specific a specific version of a text
-   * document. Whether a client supports versioned document edits is expressed via
-   * `WorkspaceClientCapabilities.versionedWorkspaceEdit`.
-   */
-  private List<TextDocumentEdit> documentChanges;
-
+public class CheWorkspaceEdit extends WorkspaceEdit {
```
Comment (tsmaeder) [inline org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: What's the reasoning for extending WorkspaceEdit? Aren' we inheriting stuff that make no sense? getResourceChanges, for example?

Comment (svor) [reply org-eclipse-che-jdt-ls-extension/jdt.ls.extension.api/src/main/java/org/eclipse/che/jdt/ls/extension/api/dto/CheWorkspaceEdit.java]: I did it to be more approximate to current model of rename command which has to return `WorkspaceEdit` object

**Pull request merged**